<a href="https://colab.research.google.com/github/borjabmadurga2-creator/Data-Analytics---Borja/blob/main/Sprint10_Borja_Belmonte.ipynb/sprint10_Borja_Belmonte.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

En este sprint, aplicaremos lo aprendido sobre APIs REST, métodos HTTP y códigos de estado, usando APIs públicas para consultar datos, filtrarlos y finalmente, guardalos en un DataFrame de pandas.

Recursos:

Nivel 1:
{JSON} Placeholder

Nivel 2:
Try Public APIs for free

Nivel 3:
Open Data BCN

**Nivel 1**

Usando la librería "requests", haremos lo siguiente:

Exploración básica con una API de laboratorio.

Consulta la API púlica JSONPlaceholder usando el método GET para obtener:

- Lista de publicaciones (/posts)
- Lista de usuarios (/users)
- Lista de tareas (/todos)


In [35]:
#importación librería requests (google colab ya la tiene instalada)
import requests
import json
import pandas as pd


In [36]:

#agrego la url a una variable para obtener la lista de publicaciones
url_posts = "https://jsonplaceholder.typicode.com/posts"

lista_publicaciones = requests.get(url_posts)
lista_publicaciones.json()



[{'userId': 1,
  'id': 1,
  'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
  'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'},
 {'userId': 1,
  'id': 2,
  'title': 'qui est esse',
  'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'},
 {'userId': 1,
  'id': 3,
  'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut',
  'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'},
 {'userId': 1,
  'id': 4,
  'title': 'eum et est occaecati',
  'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provid

In [37]:
#obtengo la lista de usuarios
url_users = "https://jsonplaceholder.typicode.com/users"

lista_users = requests.get(url_users)
lista_users.json()

[{'id': 1,
  'name': 'Leanne Graham',
  'username': 'Bret',
  'email': 'Sincere@april.biz',
  'address': {'street': 'Kulas Light',
   'suite': 'Apt. 556',
   'city': 'Gwenborough',
   'zipcode': '92998-3874',
   'geo': {'lat': '-37.3159', 'lng': '81.1496'}},
  'phone': '1-770-736-8031 x56442',
  'website': 'hildegard.org',
  'company': {'name': 'Romaguera-Crona',
   'catchPhrase': 'Multi-layered client-server neural-net',
   'bs': 'harness real-time e-markets'}},
 {'id': 2,
  'name': 'Ervin Howell',
  'username': 'Antonette',
  'email': 'Shanna@melissa.tv',
  'address': {'street': 'Victor Plains',
   'suite': 'Suite 879',
   'city': 'Wisokyburgh',
   'zipcode': '90566-7771',
   'geo': {'lat': '-43.9509', 'lng': '-34.4618'}},
  'phone': '010-692-6593 x09125',
  'website': 'anastasia.net',
  'company': {'name': 'Deckow-Crist',
   'catchPhrase': 'Proactive didactic contingency',
   'bs': 'synergize scalable supply-chains'}},
 {'id': 3,
  'name': 'Clementine Bauch',
  'username': 'Samantha

In [38]:
#obtengo la lista de tareas
url_tareas = "https://jsonplaceholder.typicode.com/todos"

lista_tareas = requests.get(url_tareas)
lista_tareas.json()

[{'userId': 1, 'id': 1, 'title': 'delectus aut autem', 'completed': False},
 {'userId': 1,
  'id': 2,
  'title': 'quis ut nam facilis et officia qui',
  'completed': False},
 {'userId': 1, 'id': 3, 'title': 'fugiat veniam minus', 'completed': False},
 {'userId': 1, 'id': 4, 'title': 'et porro tempora', 'completed': True},
 {'userId': 1,
  'id': 5,
  'title': 'laboriosam mollitia et enim quasi adipisci quia provident illum',
  'completed': False},
 {'userId': 1,
  'id': 6,
  'title': 'qui ullam ratione quibusdam voluptatem quia omnis',
  'completed': False},
 {'userId': 1,
  'id': 7,
  'title': 'illo expedita consequatur quia in',
  'completed': False},
 {'userId': 1,
  'id': 8,
  'title': 'quo adipisci enim quam ut ab',
  'completed': True},
 {'userId': 1,
  'id': 9,
  'title': 'molestiae perspiciatis ipsa',
  'completed': False},
 {'userId': 1,
  'id': 10,
  'title': 'illo est ratione doloremque quia maiores aut',
  'completed': True},
 {'userId': 1,
  'id': 11,
  'title': 'vero rerum

También se podría hacer todo en una sola celda guardando los resultados en una variable e imprimir solo 2 ejemplos de cada

In [39]:
#solución e nuna misma celda

url_publicaciones = "https://jsonplaceholder.typicode.com/posts"

lista_posts = requests.get(url_publicaciones)
publicaciones = lista_posts.json()

url_usuarios = "https://jsonplaceholder.typicode.com/users"
lista_usuarios = requests.get(url_users)
users = lista_usuarios.json()

url_tareas = "https://jsonplaceholder.typicode.com/todos"
lista_tasks = requests.get(url_tareas)
tareas = lista_tasks.json()

print("primeras 2 publicaciones: ")
print(publicaciones[:2])
print("primeros 2 usuarios: ")
print(users[:2])
print("primeras 2 tasks: ")
print(tareas[:2])

primeras 2 publicaciones: 
[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}]
primeros 2 usuarios: 
[{'id': 1, 'name': 'Leanne Graham', 'username': 'Bret', 'email': 'Sincere@april.biz', 'address': {'street': 'Kulas Light', 'suite': 'Apt. 556', 'city': 'Gwenborough', 'zipcode': '92998-3874', 'geo': {'lat': '-37.3159', 'lng': '81.1496'}}, 'phone': '1-770-736-8031 x56442', 'website': 'hildegard.org', 'company': {'name': 'Romaguera-Crona', 'catchPhrase': 'Multi-layered client-server neural-net',

Muestra por pantalla:

- La cantidad total de cada recurso
- El código de estado de cada petición

In [40]:
#Al tratarse de una lista de Python, se puede usar la función integrada len() para saber el total de cada recurso
publicaciones = lista_publicaciones.json()
usuarios = lista_users.json()
tasks = lista_tareas.json()
print(f"La cantidad de publicaciones total son: {len(publicaciones)}")
print(f"La cantidad de usuarios totales son: {len(usuarios)}")
print(f"la cantidad de tareas totales son: {len(tasks)}")

La cantidad de publicaciones total son: 100
La cantidad de usuarios totales son: 10
la cantidad de tareas totales son: 200


In [41]:
#muestro el código de estado de cada petición
publi = lista_publicaciones.status_code
usu = lista_users.status_code
tar = lista_tareas.status_code
print(f"El código de estado de la petición de publicaciones es: {publi} ")
print(f"El código de estado de la petición de usuarios es: {usu} ")
print(f"El código de estado de la petición de tareas es: {tar}")

El código de estado de la petición de publicaciones es: 200 
El código de estado de la petición de usuarios es: 200 
El código de estado de la petición de tareas es: 200


El código 200 muestra que la solicitud fue reconocida por el servidor, aceptada y está siendo procesada

Haz una petición a una publicación inexistente para obtener un error 404 y muestre el código de estado recibido

In [42]:
#petición a una publicación inválida
post_url = ("https://jsonplaceholder.typicode.com/posts/300")
response = requests.get(post_url)
print(response)

<Response [404]>


En este caso, como sabemos que el recurso  /posts tiene 100 registros, en la url, debemos apuntar directamente al identificador que queremos que nos muestre por pantalla. En este caso, hemos pasado el 300 (inexistente), es por ello que arroja el error 404 "not found", la url no existe.

Haz una petición POST para crear una nueva publicación ficticia. Incluye el título, el cuerpo del mensaje y un userId.

- Muestra la respuesta JSON
- Muestra el código de estado


In [43]:
#creación de una nueva publicación ficticia
publi_url = "https://jsonplaceholder.typicode.com/posts/"

datos = {#añado los datos que quiero añadir a la publicación en formato diccionario clave : valor
    "title" : "my first post",
    "body" : "Hello from Python!",
    "userId" : 12
}

response = requests.post(publi_url, json=datos) #paso el resquests POST y serializo los datos con json
print(f"Respuesta JSON: {response.json()}")#muestro por pantalla los resultados en formato json
print(f"Estatus: {response.status_code}")

Respuesta JSON: {'title': 'my first post', 'body': 'Hello from Python!', 'userId': 12, 'id': 101}
Estatus: 201


Añado los datos ficticios que quiero introducir mediante un diccionario de Python (clave : valor). Paso el requests POST, la url y los datos creados en formato JSON y muestro por pantalla.
El resultado del código de estado obtenido es 201. La serie de códigos 200 indican que la solicitud fue reconocida por el servidor, aceptada y está siendo procesada.
El código 201 significa "creado" La solicitud se cumplió y el servidor creó un nuevo recurso.

Haz una petiicón PATCH para modificar parcialmente una publicación existente.

- Muestra la respuesta JSON
- Muestra el código de estado

In [44]:
#En este caso, vamos a modificar parcialmente una publicación existente, concretamente, el titulo
modificacion_url = "https://jsonplaceholder.typicode.com/posts/1"#uso la url apuntando al id 1 (primer publicación)

datos_modificados = {
    "title" : "my post has been modified"#le digo qué datos quiero modificar, en este caso el titulo y le digo qué quiero poner en el diccionario python
}

response = requests.patch(modificacion_url, json=datos_modificados)#uso el requests patch para modificar parcialmente la entrada, le paso la url y serializo los datos
print(f"Respuesta JSON: {response.json()}")#muestro los datos
print(f"Estatus: {response.status_code}")

Respuesta JSON: {'userId': 1, 'id': 1, 'title': 'my post has been modified', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}
Estatus: 200


Haz una petición DELETE sobre una publicación.

- Muestra la respuesta JSON
- Muestra el código de estado

In [45]:
#en este caso, vamos a eliminar la quinta publicación
eliminacion_url = "https://jsonplaceholder.typicode.com/posts/5"#apunto a la quinta posición

response = requests.delete(eliminacion_url)
print(f"Respuesta JSON: {response.json()}")
print(f"Estatus: {response.status_code}")

Respuesta JSON: {}
Estatus: 200


Vemos que la respuesta es un diccionario vacío, ya que hemos eliminado la publicación número 5. El estatus que muestra es 200 (procesada). En ocasiones también puede sacar estatus 204 "no content", puesto que ha sido eliminada.

**Nivel 2**

  Explora el repositorio de Public APIs y escoge una API que permita hacer peticiones GET.

In [46]:
url_api1 = "https://blockchain.info/latestblock"

response1 = requests.get(url_api1)
response1.json()

{'hash': '0000000000000000000040898fe3efc9d6d6b27de1fbe3b231ecf7046293faca',
 'time': 1787848386,
 'block_index': 964323,
 'height': 964323,
 'txIndexes': [695966919965635,
  6751179972774911,
  4827173109721259,
  7992044013347608,
  7594700067538972,
  7448817398954901,
  6933421418703247,
  5098997404279742,
  6919830440550699,
  428361553049544,
  6233788072812035,
  7234742418267624,
  8585076892542027,
  1894045886367380,
  4769707225193197,
  6107024861289824,
  6178758285367815,
  7854184825337085,
  7714472476716739,
  7487978502062666,
  4111707145621677,
  7500448249834899,
  581618014090477,
  6926204237876521,
  4009407251492127,
  5751530883730962,
  6858495539073548,
  5694646600719286,
  5419271489798143,
  7536190193341809,
  4933047380716701,
  5622253395084272,
  1199960890288063,
  8304132903334764,
  8142211148132561,
  3930937494324961,
  4305448594716682,
  2474824860630538,
  2149150629151313,
  2232248005858890,
  2389579487510361,
  3855520033775614,
  4552918

In [47]:
url_api2 = "https://blockchain.info/unconfirmed-transactions?format=json"

response2 = requests.get(url_api2)
response2.json()

{'txs': [{'hash': 'bf9e410013a4ffc8a9e0bbe95f2be1d92ab1831df37b003d92fa6a053b8d80e4',
   'ver': 1,
   'vin_sz': 1,
   'vout_sz': 2,
   'size': 222,
   'weight': 561,
   'fee': 278,
   'relayed_by': '0.0.0.0',
   'lock_time': 0,
   'tx_index': 8039704844873055,
   'double_spend': False,
   'time': 1787848623,
   'block_index': None,
   'block_height': None,
   'inputs': [{'sequence': 4294967293,
     'witness': '0247304402200e5e6ed951771b35f1421b148b77dba2de10fa117386203ece6befc10b819ad70220272d2327481455f842c59067bea0ac23f3ef1d339cd4eff242eb83d06907e3440121034013c567349250bf27450a870c24c47078fcd8b11512c05c2bb8774bc99373bb',
     'script': '',
     'index': 0,
     'prev_out': {'type': 0,
      'spent': True,
      'value': 29619,
      'spending_outpoints': [{'tx_index': 8039704844873055, 'n': 0}],
      'n': 0,
      'tx_index': 6920284783195131,
      'script': '0014bfb0ceed3d20c419befd24fab974b5f181313ca5',
      'addr': 'bc1qh7cvamfayrzpn0haynatja947xqnz099el89hu'}}],
   'out': [{'

Lee la documentación de la API:

- Revisa la sección de endpoints disponibles. En un markdown, apunta como mínimo a dos enpoints diferents.

- Comprueba si ofrecen filtros o parámetros opcionales interesantes. Anótalo en un markdown.

- Verifica que la respuesta sea en formato JSON (es el más adecuado para este ejercicio)


**El primer endpoint es**: "https://blockchain.info/latestblock"
que muestra el bloque más reciente validado y añadido de forma permanente a una cadena de bloques.
Los campos principales son 'hash', 'time', 'block_index', 'height' y 'txIndexes'

**El segundo endpoint es**: "https://blockchain.info/unconfirmed-transactions?format=json"
que muestra un diccionario que contiene una lista de transacciones ('txs'), donde cada transacción es a su vez un diccionario.
Los campos principales son 'hash' y una serie de datos relacionados con las transacciones ('size','weight','fee','time') y otros diccionarios dentro.

Se confirma que la respuesta es JSON, ya que ha sido posible usar el método .json() de la libreria requests sin errores, obteniendo un diccionario de Python para su lectura.

Haz una petición GET sencilla:

- Muestra el código de estado de la respuesta
- Imprime de forma clara algnos de los campos de respuesta JSON

In [48]:
#conocer el código del estado de la consulta. para hacerlo más fácil y ayudarme en el siguiente paso, los guardaré en variables
datos1 = response1.status_code
datos2 = response2.status_code
print(f"El estado de código de la petición 1 es: {datos1}")
print(f"El estado de código de la petición 2 es: {datos2}")

El estado de código de la petición 1 es: 200
El estado de código de la petición 2 es: 200


In [49]:
#imprimo algunos de los campos de respuesta JSON
datos1 = response1.json()
hash = datos1["hash"]
time = datos1["time"]

print(f"El hash es: {hash}")
print(f"El tiempo es: {time}")


El hash es: 0000000000000000000040898fe3efc9d6d6b27de1fbe3b231ecf7046293faca
El tiempo es: 1787848386


In [50]:
#imprimo algunos de los campos de respuesta JSON
datos2 = response2.json()
hash = datos2["txs"][0]["hash"]
fee = datos2["txs"][0]["fee"]
print(f"El hash es: {hash}")
print(f"El peso es: {fee}")


El hash es: bf9e410013a4ffc8a9e0bbe95f2be1d92ab1831df37b003d92fa6a053b8d80e4
El peso es: 278


Convierte los datos a un DataFrame de pandas:

- Muestra las primera filas del Dataframe

In [51]:
#para pasarlo a dataframe, envuelvo la variable entre corchetes precisamente por que es una lista con un único diccionario dentro. Solo muestra una fila por que solo almacena el último registro
df_datos1 = pd.DataFrame([datos1])
df_datos1

,hash,time,block_index,height,txIndexes
0,0000000000000000000040898fe3efc9d6d6b27de1fbe3...,1787848386,964323,964323,"[695966919965635, 6751179972774911, 4827173109..."


In [52]:
#en este caso, al tratarse de un diccionario que contiene una lista de transacciones bajo la clave txs. Para pasarloa dataframe accedo directamaente a datos2["txs"] que ya es una lista de diccionarios lista para usar
df_datos2 = pd.DataFrame(datos2["txs"])
df_datos2

,hash,ver,vin_sz,vout_sz,size,weight,fee,relayed_by,lock_time,tx_index,double_spend,time,block_index,block_height,inputs,out,rbf
0,bf9e410013a4ffc8a9e0bbe95f2be1d92ab1831df37b00...,1,1,2,222,561,278,0.0.0.0,0,8039704844873055,False,1787848623,None,None,"[{'sequence': 4294967293, 'witness': '02473044...","[{'type': 0, 'spent': False, 'value': 683, 'sp...",True
1,94268a34bb0c777b226952d7984549cd93fd1a32324b4f...,2,1,2,225,573,322,0.0.0.0,0,7568518019680362,False,1787848623,None,None,"[{'sequence': 4294967293, 'witness': '02473044...","[{'type': 0, 'spent': False, 'value': 477683, ...",True
2,b4652f5f66328ce6d4b24c18f020016d58bba93c1fb3f0...,1,11,2,1708,3277,3412,0.0.0.0,0,5762267476876206,False,1787848623,None,None,"[{'sequence': 4294967295, 'witness': '02483045...","[{'type': 0, 'spent': False, 'value': 30000000...",NaN
3,b1b3cb1d9b64a7b53817a72fabf11c7446d42d01cae86b...,2,1,2,222,558,477,0.0.0.0,0,1922598812794926,False,1787848623,None,None,"[{'sequence': 4294967293, 'witness': '02483045...","[{'type': 0, 'spent': False, 'value': 6227, 's...",True
4,4c58fbe6a56f1acaef1b2ecc93432d7d1b17a686713506...,2,2,1,339,1356,1690,0.0.0.0,0,1793222960326223,False,1787848623,None,None,"[{'sequence': 4294967293, 'witness': '', 'scri...","[{'type': 0, 'spent': False, 'value': 6457168,...",True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,53febc49abb2b0bd0aa910ac4fac164d98076815db54bd...,1,3,2,563,1187,365,0.0.0.0,0,7857916614956073,False,1787848602,None,None,"[{'sequence': 4294967295, 'witness': '0141d729...","[{'type': 0, 'spent': False, 'value': 37355, '...",NaN
96,bcf6df9758cefed00e6727526f922c6794fc76bc3ba7be...,2,2,1,344,722,183,0.0.0.0,0,7176602837585693,False,1787848602,None,None,"[{'sequence': 4294967293, 'witness': '02483045...","[{'type': 0, 'spent': False, 'value': 58346, '...",True
97,277538cacfec0f75722b6943af713618566f8deed0bfda...,2,1,2,221,557,443,0.0.0.0,0,5868297410821625,False,1787848602,None,None,"[{'sequence': 4294967293, 'witness': '02473044...","[{'type': 0, 'spent': False, 'value': 9594, 's...",True
98,30e5d9a61ace60f00e9add4218aa1ce263da06fe138abc...,2,1,2,414,792,318,0.0.0.0,964321,5114369335584356,False,1787848602,None,None,"[{'sequence': 4294967293, 'witness': '04004730...","[{'type': 0, 'spent': False, 'value': 5975611,...",True


In [53]:
#para desanidar un nivel más, cada transacción tiene una lista de  "inputs". Con meta indico que los campos hash fee y time pertenecen al primer nivel (txs), y marco la ruta completa ["txs","hash"] para mostrar esas columnas de primer nivel
df_datos2 = pd.json_normalize(
    datos2,
    record_path=["txs","inputs"],
    meta=[['txs','hash'],['txs','fee'],['txs','time']]
)
df_datos2

,sequence,witness,script,index,prev_out.type,prev_out.spent,prev_out.value,prev_out.spending_outpoints,prev_out.n,prev_out.tx_index,prev_out.script,prev_out.addr,txs.hash,txs.fee,txs.time
0,4294967293,0247304402200e5e6ed951771b35f1421b148b77dba2de...,,0,0,True,29619,"[{'tx_index': 8039704844873055, 'n': 0}]",0,6920284783195131,0014bfb0ceed3d20c419befd24fab974b5f181313ca5,bc1qh7cvamfayrzpn0haynatja947xqnz099el89hu,bf9e410013a4ffc8a9e0bbe95f2be1d92ab1831df37b00...,278,1787848623
1,4294967293,024730440220751faf9a667df1b4692075c32d274ca906...,,0,0,True,521663,"[{'tx_index': 7568518019680362, 'n': 0}]",0,1232654863526218,00147ef1896fdf6f11a781e583a7f380e2e43fd980f8,bc1q0mccjm7ldug60q09swnl8q8zuslanq8cxp2gaa,94268a34bb0c777b226952d7984549cd93fd1a32324b4f...,322,1787848623
2,4294967295,02483045022100e3e6386e17fb88faea94968ef9ce60b1...,,0,0,True,30000000,"[{'tx_index': 5762267476876206, 'n': 0}]",0,5574514746435713,00147a498404b05f69d56caecdea4acf07e2a9efc621,bc1q0fycgp9sta5a2m9weh4y4nc8u257l33p5tzykc,b4652f5f66328ce6d4b24c18f020016d58bba93c1fb3f0...,3412,1787848623
3,4294967295,0247304402200e41ca12072c7123c06aacf0043f2bb037...,,1,0,True,14806900,"[{'tx_index': 5762267476876206, 'n': 1}]",0,3971237944223387,00147a498404b05f69d56caecdea4acf07e2a9efc621,bc1q0fycgp9sta5a2m9weh4y4nc8u257l33p5tzykc,b4652f5f66328ce6d4b24c18f020016d58bba93c1fb3f0...,3412,1787848623
4,4294967295,02483045022100bd165becd9e576a3e9bd4608ef4d5d36...,,2,0,True,6108882,"[{'tx_index': 5762267476876206, 'n': 2}]",0,6686152831020606,00147a498404b05f69d56caecdea4acf07e2a9efc621,bc1q0fycgp9sta5a2m9weh4y4nc8u257l33p5tzykc,b4652f5f66328ce6d4b24c18f020016d58bba93c1fb3f0...,3412,1787848623
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
205,4294967293,02483045022100b826d4fd2c3d57d37371618df84c3eb0...,,0,0,True,30774,"[{'tx_index': 7176602837585693, 'n': 0}]",2,7471125184489629,00141f091b53289ab5ae5db3e382c2b74a48d89f62e0,bc1qruy3k5egn266uhdnuwpv9d62frvf7chqplpnsh,bcf6df9758cefed00e6727526f922c6794fc76bc3ba7be...,183,1787848602
206,4294967293,02483045022100cd5d5d48f184c0ce07205f468e301f05...,,1,0,True,27755,"[{'tx_index': 7176602837585693, 'n': 1}]",18,8380104736770842,00141f091b53289ab5ae5db3e382c2b74a48d89f62e0,bc1qruy3k5egn266uhdnuwpv9d62frvf7chqplpnsh,bcf6df9758cefed00e6727526f922c6794fc76bc3ba7be...,183,1787848602
207,4294967293,02473044022008086cd7c732e0876b1beaef0d549cf0de...,,0,0,True,10037,"[{'tx_index': 5868297410821625, 'n': 0}]",0,5473940507070030,00145598af355466d36e6a671374abbb2b75ef477956,bc1q2kv27d25vmfku6n8zd62hwetwhh5w72kxpw8cd,277538cacfec0f75722b6943af713618566f8deed0bfda...,443,1787848602
208,4294967293,040047304402203a2766ac37cc1a4550c81bd19a7239d8...,,0,0,True,18396500,"[{'tx_index': 5114369335584356, 'n': 0}]",1,1314272008288201,00209ab4659824e92de01bddb37db69d09b4fa009024f7...,bc1qn26xtxpyayk7qx7akd7md8gfknaqpypy7u8826cqw8...,30e5d9a61ace60f00e9add4218aa1ce263da06fe138abc...,318,1787848602


**Nivel 3**

Utiliza la API de Open BCN para:

- Encontrar un DataSet de tu interés mediante package_search y package_show


In [54]:
#el primer endpoint es con package_search y paso una string para que busque la palabra economia (uso la q con package_search)
url_base = "https://opendata-ajuntament.barcelona.cat/data/api/action/package_search?q=economia"
respuesta = requests.get(url_base)
datos = respuesta.json()
datasets = datos["result"]["results"]
print(len(datasets))


10


En la siguiente celda, hago un bucle anidado for, con un condicional if para que muestre los ids de los datasets que tengan formato csv

In [55]:
for dataset in datasets:
    #print(dataset['title'])
    #print(dataset["id"])
    print(f"{dataset["title"]} - {dataset["id"]}")
    for recurso in dataset["resources"]:
      if recurso["format"] == "CSV":
        print(recurso["id"])

Fees excluding VAT from the public price register for waste collection (excluding large producers) of the Barcelona City Council - 331cba6c-7ca7-49d7-8ec7-19ff39d70aa3
806839e7-6798-48c9-bf04-caf045c959c1
fb3d9998-3d60-407c-8654-948b5603862c
Receipts and quota of the register of the Vados Tax of the Barcelona City Council - 7e63eeb0-670a-4bcc-bee9-1d89697a4501
e35dce7d-96f6-4130-b794-42967da2cce5
5f5e606d-4a38-4b4a-a16a-38e024e0b642
Receipts and charges from the motor vehicle tax roll of the Barcelona City Council - e21b436f-7f84-4f7b-a3de-f7dc6b43264f
110749d9-ee3e-4d1f-b962-680e3cebe608
Receipts and quota of the register of the Fee for private use of the public thoroughfare (nightstands and terraces) of the Barcelona City Council - f725406c-9c7a-4ed1-86cb-3358d6fb4dd0
7c8e7078-c52e-40f6-9123-af196387e057
da13ea98-1a06-4eca-9dd0-36848c3f2b54
b821a0e7-eb15-4c89-bae8-f8f21f1996b6
Vehicle types in traffic complaints and sanctions processed by the Institut Municipal d’Hisenda of Barcelona

De los resultados obtenidos, selecciona uno que tenga recursos disponibles en CSV o JSON


In [56]:
#el segundo endpoint es con packache_show y le paso el id de un dataset con recursos disponibles en csv
url_show = "https://opendata-ajuntament.barcelona.cat/data/api/action/package_show?id=331cba6c-7ca7-49d7-8ec7-19ff39d70aa3"
respuesta_show = requests.get(url_show)
print(json.dumps(respuesta_show.json(), indent=4, ensure_ascii=False))

{
    "help": "https://opendata-ajuntament.barcelona.cat/data/api/3/action/help_show?name=package_show",
    "success": true,
    "result": {
        "notes_translated": {
            "ca": "Quotes sense IVA del padró del preu públic de recollida de residus (sense grans productors ) de l’Ajuntament de Barcelona per districte, tipus de productor, secció, epígraf i classificació de la fracció residual generada. No inclou les liquidacions emeses posteriors a la publicació del padró.",
            "en": "Fees excluding VAT from the public price register for waste collection (excluding large producers) of the Barcelona City Council, by district, producer type, section, heading (tariff item) and classification. It does not include settlements made after the publication of the register.",
            "es": "Cuotas sin IVA del padrón del precio público de recogida de residuos (sin grandes productores) del Ayuntamiento de Barcelona por distrito, tipo de productor, sección, epígrafe y clasificac

Con el resource_id del recurso seleccionado, realiza una consulta mediante datastore_search para:

- Recuperar, al menos, 100 registros

In [57]:
#utilizo el endpoint datastore_search, con uno de los ids obtenido anteriormente que tengan recursos en csv o json
url_datastore = "https://opendata-ajuntament.barcelona.cat/data/api/action/datastore_search?resource_id=806839e7-6798-48c9-bf04-caf045c959c1"
respuesta_datastore = requests.get(url_datastore)
datos_ds = respuesta_datastore.json()
print(len(datos_ds['result']['records']))


100


Convertir los resultados en DataFrame

In [58]:
#conversion a dataframe
df = pd.json_normalize(datos_ds, record_path=['result','records'])#uso json_normalize para aplanar un json anidado, el record_path es para indicarle la ruta donde debe buscar, en result y dentro, record
df.head()

,Descripcio_Fraccio_Residual_Generada,Descripcio_Epigraf,Codi_Districte,Nom_Districte,Tipus_Productor,Codi_Epigraf,Descripcio_Seccio,Total_Quota_Sense_IVA,Codi_Seccio,Codi_Fraccio_Residual _Generada,_id
0,Resta,Arquitectes,6,Gràcia,MÍNIM,411,Activitats Professionals,1617,2,A,1
1,"Resta, paper-cartró, vidre i envasos",Altres serveis de cafeteries i bars,2,Eixample,MITJÀ,6732,Activitats Empresarials,299449.55,1,C,2
2,"Resta, paper-cartró, vidre i envasos",Altres serveis de cafeteries i bars,7,Horta-Guinardó,MITJÀ,6732,Activitats Empresarials,62042.09,1,C,3
3,Resta i paper-cartró,Comerç al detall de béns de segona mà,2,Eixample,MITJÀ,656,Activitats Empresarials,6040.99,1,B,4
4,"Resta, orgànica, paper-cartró, vidre i envasos",Serveis de restaurants d'una forquilla,1,Ciutat Vella,MITJÀ,6715,Activitats Empresarials,158332.64,1,E,5


Guardar el DataFrame en un archivo csv

In [59]:
df.to_csv("mis_resultados.csv", index=False)#to_csv para convertirlo. lo guarda en el entorno local de colab, no en el ordenador. Le quito el index para que no lo ponga como columna